### PR TITLE
feat: add parsing of effect declarations

### PR DESCRIPTION
**Pull request by mlutze**
related to #3542, depends on #3544

### CODE REVIEW

File: main/test/flix/Test.Dec.Effect.flix [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+        def readLine(): String
+        def println(s: String): Unit
+    }
+
+    eff Empty
+
+    // Will be removed after Weeder checks are added
+    eff Polymorphic[a] {
+        def op[a](): a
+    }
```
Comment (magnus-madsen) [inline main/test/flix/Test.Dec.Effect.flix:13]: I think you can already a bunch more of tests :)

Comment (mlutze) [reply main/test/flix/Test.Dec.Effect.flix:13]: Were there other particular vectors of possible error you were looking for, or should I just reproduce a bunch of the standard effects from literature?

Comment (magnus-madsen) [reply main/test/flix/Test.Dec.Effect.flix:13]: Not specifically, just some systematic testing.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 2]
```diff
@@ -112,6 +112,22 @@ object ParsedAst {
...
+      * @param mod        the associated modifiers.
+      * @param sp1        the position of the first character in the declaration.
+      * @param ident      the name of the definition.
+      * @param tparams    the type parameters.
+      * @param fparamsOpt the formal parameters.
+      * @param tpe        the declared type.
+      * @param tconstrs   the type constraints.
+      * @param sp2        the position of the last character in the declaration.
+      */
+    case class Op(doc: ParsedAst.Doc, ann: Seq[ParsedAst.Annotation], mod: Seq[ParsedAst.Modifier], sp1: SourcePosition, ident: Name.Ident, tparams: ParsedAst.TypeParams, fparamsOpt: Seq[ParsedAst.FormalParam], tpe: ParsedAst.Type, pur: Option[ParsedAst.Type], tconstrs: Seq[ParsedAst.TypeConstraint], sp2: SourcePosition) extends ParsedAst.Declaration
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:129]: - Do we want to call this `Op` or `Operation`?

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:129]: I went with `Op` to match `Sig` and `Def`

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 3]
```diff
@@ -237,6 +242,24 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
...
+      def Head = rule {
+        Documentation ~ Modifiers ~ SP ~ keyword("eff") ~ WS ~ Names.Effect ~ optWS ~ TypeParams
+      }
+
+      def EmptyBody = namedRule("EffectBody") {
+        push(Nil) ~ SP
+      }
+
+      def NonEmptyBody = namedRule("EffectBody") {
+        optWS ~ "{" ~ zeroOrMore(Declarations.Op) ~ optWS ~ "}" ~ SP
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:255]: Why space on the other side of `{`?

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 4]
```diff
@@ -237,6 +242,24 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
       }
     }
 
+    def Effect: Rule1[ParsedAst.Declaration] = {
+      def Head = rule {
+        Documentation ~ Modifiers ~ SP ~ keyword("eff") ~ WS ~ Names.Effect ~ optWS ~ TypeParams
+      }
+
+      def EmptyBody = namedRule("EffectBody") {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:250]: I thought we were using these namedRule sparingly?

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 5]
```diff
@@ -237,6 +242,24 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
       }
     }
 
+    def Effect: Rule1[ParsedAst.Declaration] = {
+      def Head = rule {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:246]: Dont like the name Head, but I see we use it elsewhere, so okay.

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:255]: Whitespace*

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:129]: Ok. Agreed!

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 6]
```diff
@@ -116,7 +116,11 @@ object Weeder {
 
     case d: ParsedAst.Declaration.Instance => visitInstance(d)
 
+    case _: ParsedAst.Declaration.Effect => Nil.toSuccess // ignoring effect declarations for now
+
     case _: ParsedAst.Declaration.Sig => throw InternalCompilerException(s"Unexpected declaration")
+
+    case _: ParsedAst.Declaration.Op => throw InternalCompilerException(s"Unexpected declaration")
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:123]: Does this crash the compiler or can it not happen?

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala (hunk 3 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:255]: This is a parallel to what's done in classes and instances. I'll try removing it and see what happens.

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala (hunk 4 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:250]: Parallels Instance and Class.

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala (hunk 6 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:123]: Can't happen. This visitor is at the top level, where the parser will never allow an operation.

Maybe we can remove Sig and Op from the Declaration enum.

**Pull request merged**